server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java

```diff
@@ -380,6 +418,74 @@ private static boolean assertShardStateConsistent(
         return true;
     }
 
+    /**
+     * Adds any new node IDs to {@link #nodesIdsForRemoval}, and removes any node IDs that are no longer marked for shutdown if they have no
+     * running shard snapshots.
+     */
+    public SnapshotsInProgress withUpdatedNodeIdsForRemoval(ClusterState clusterState) {
+        final var updatedNodeIdsForRemoval = new HashSet<>(nodesIdsForRemoval);
+
+        final var nodeIdsMarkedForRemoval = getNodesIdsMarkedForRemoval(clusterState);
+
+        // add any nodes newly marked for removal
+        updatedNodeIdsForRemoval.addAll(nodeIdsMarkedForRemoval);
+
+        // remove any nodes which are no longer marked for shutdown if they have no running shard snapshots
+        updatedNodeIdsForRemoval.removeAll(getObsoleteNodeIdsForRemoval(nodeIdsMarkedForRemoval));
+
+        if (updatedNodeIdsForRemoval.equals(nodesIdsForRemoval)) {
+            return this;
+        } else {
+            return new SnapshotsInProgress(entries, Collections.unmodifiableSet(updatedNodeIdsForRemoval));
+        }
+    }
+
+    private static Set<String> getNodesIdsMarkedForRemoval(ClusterState clusterState) {
+        final var nodesShutdownMetadata = clusterState.metadata().nodeShutdowns();
+        final var shutdownMetadataCount = nodesShutdownMetadata.getAllNodeIds().size();
+        if (shutdownMetadataCount == 0) {
+            return Set.of();
+        }
+
+        final Set<String> result = Sets.newHashSetWithExpectedSize(shutdownMetadataCount);
+        for (final var entry : nodesShutdownMetadata.getAll().entrySet()) {
+            if (entry.getValue().getType() != SingleNodeShutdownMetadata.Type.RESTART) {
+                // Only pause the snapshot when the node is being removed (to let shards vacate) and not when it is restarting in place. If
+                // it is restarting and there are replicas to promote then we need #71333 to move the shard snapshot over; if there are no
```

Comment:
My concern was that we do not vacate a node marked for `RESTART` (nor do we wait for it to be vacated) so pausing the snapshot of its shards would delay snapshotting until the node restarts, and there's nothing we can do about that from within Elasticsearch so it might be quite a delay.

But also the implementation as presented here doesn't work for the `RESTART` case because we do not move shard snapshots from `WAITING` to `INIT` when a replica is promoted to primary.

Reply:
> But also the implementation as presented here doesn't work for the RESTART case because we do not move shard snapshots from WAITING to INIT when a replica is promoted to primary.

Why is this not happening? IIUC, the node assignment for a shard is computed with `SnapshotsService#shards` and more specifically in [initShardSnapshotStatus](https://github.com/elastic/elasticsearch/pull/101717/files#diff-de99cbad875eb82467666e03b94c5b9376e92ce73fa021a99ab58ea4c2bae4f6R2878). It seems to me whenever primary becomes available the correspond shard snapshot will be moved to `INIT`?

Reply:
Hmm, I thought we skipped this case in `waitingShardsStartedOrUnassigned` but now that I look again I think it might be fine.